metrics/metricstest/metricstest.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricstest
+
+import (
+	"testing"
+
+	"go.opencensus.io/stats/view"
+)
+
+// CheckStatsReported checks that there is a view registered with the given name for each string in names,
+// and that each view has at least one record.
+func CheckStatsReported(t *testing.T, names ...string) {
+	t.Helper()
+	for _, name := range names {
+		d, err := view.RetrieveData(name)
+		if err != nil {
+			t.Errorf("Reporter.Report() error = %v, wantErr %v", err, false)
+		}
+		if len(d) < 1 {
+			t.Error("No data reported when data was expected, view data is empty.")
+		}
+	}
+}
+
+// CheckStatsNotReported checks that there are no records for any views that a name matching a string in names.
+// Names that do not match registered views are considered not reported.
+func CheckStatsNotReported(t *testing.T, names ...string) {
+	t.Helper()
+	for _, name := range names {
+		if d, err := view.RetrieveData(name); err != nil {
+			if len(d) > 0 {
+				t.Errorf("Unexpected data reported when no data was expected. Reporter len(d) = %d", len(d))
+			}
+		}
+	}
+}
+
+// CheckCountData checks the view with a name matching string name to verify that the CountData stats
+// reported are tagged with the tags in wantTags and that wantValue matches reported count.
+func CheckCountData(t *testing.T, name string, wantTags map[string]string, wantValue int64) {
+	t.Helper()
+	if row := checkExactlyOneRow(t, name); row != nil {
+		for _, got := range row.Tags {
+			n := got.Key.Name()
+			if want, ok := wantTags[n]; !ok {
+				t.Errorf("Reporter got an extra tag %v: %v", n, got.Value)
+			} else if got.Value != want {
+				t.Errorf("Reporter expected a different tag value for key: %s, got: %s, want: %s", n, got.Value, want)
+			}
+		}
```

Comment:
You might get a clearer output if you convert row to a map[string]string, and then use [`cmp.Diff`](https://godoc.org/github.com/google/go-cmp/cmp#Diff) with an [`cmpopts.IgnoreMapEntries`](https://godoc.org/github.com/google/go-cmp/cmp/cmpopts#IgnoreMapEntries) argument.

Reply:
Interesting, I will give it a try.

Reply:
I'm going to leave it as-is because the error messages here are clear enough. Using cmp.Diff with cmpopts.IgnoreMapEntries is a little bit more complex code-wise and more inefficient algorithmically.